src/rapids_pre_commit_hooks/pyproject_license.py

```diff
@@ -0,0 +1,71 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import random
+
+import tomlkit
+import tomlkit.exceptions
+
+from .lint import LintMain
+
+RAPIDS_LICENSE = "Apache 2.0"
+
+
+def find_value_location(document, key):
+    copied_document = copy.deepcopy(document)
+    node = copied_document
+    while str(placeholder := random.randint(0, 1048576)) in node.as_string():
+        pass
+
+    while len(key) > 1:
+        node = node[key[0]]
+        key = key[1:]
+    node[key[0]] = placeholder
+
+    begin_loc = copied_document.as_string().find(str(placeholder))
+    end_loc = (
+        begin_loc
+        + len(str(placeholder))
+        - len(copied_document.as_string())
```

Comment:
I don't know how fast this function is, but you already compute it once above so you might as well cache `copied_document.as_string()` so that you can both evaluate its length and find the placeholder without needing to call it again.

Reply:
Actually, better idea. Since you are guaranteed that your placeholder is a unique string that does not exist in the file, why not skip the traversal of the TOML hierarchy for replacement and instead just do (roughly, this skips some steps)
```
doc_str = document.as_string()
copied = copy(doc_str)  # At this point it's just a string, so copying will be faster
original_val = document.get(...)  # This is pseudocode, you'll have to do the recursive traversal you showed above to get the first arg
copied.replace(original_val, placeholder)
begin_val = copied.find(str(placeholder))
end_loc = begin_loc + len(original_val)
```

Reply:
Is `copied` supposed to be a string or a parsed TOML document?

If it's a string, replacing `original_val` with `placeholder` does not guard against `original_val` appearing multiple times, hence the reason for the placeholder in the first place. Otherwise, we might as well just do a search for `original_val`.

Reply:
That's fair, I didn't consider the possibility of the license appearing twice but it's certainly possible. In that case you're right that you need to do the replacement on the TOML document before you convert it to a string, and I revert to my original, simpler, suggestion: you should be able to only call `copied_document.as_string()` once. Also I still don't think you shouldn't need these three `len`s since IIUC you're basically computing the length of the original value, which you can compute after the recursive traversal of nodes before you assign placeholder.